File: cdm-sources/src/cdm/event/common/functions/ResolveTransfer.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.event.common.functions.InterestCashSettlementAmount import InterestCashSettlementAmount
from cdm.event.common.CalculateTransferInstruction import CalculateTransferInstruction
from cdm.event.common.functions.EquityCashSettlementAmount import EquityCashSettlementAmount
from cdm.event.common.Transfer import Transfer
from cdm.event.common.functions.SecurityFinanceCashSettlementAmount import SecurityFinanceCashSettlementAmount

__all__ = ['ResolveTransfer']


@replaceable
def ResolveTransfer(instruction: CalculateTransferInstruction) -> Transfer:
    """
    Defines how to calculate the amount due to be transferred after a Reset Event.
    
    Parameters 
    ----------
    instruction : CalculateTransferInstruction
    
    Returns
    -------
    transfer : Transfer
    
    """
    self = inspect.currentframe()
    
    
    def _then_fn2():
        return InterestCashSettlementAmount(rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "tradeState"), rosetta_resolve_attr(self, "payout"), rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "resets"), rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "date"))
    
    def _else_fn2():
        return True
    
    def _then_fn1():
        return EquityCashSettlementAmount(rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "tradeState"), rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "date"))
    
    def _else_fn1():
        return if_cond_fn((rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "payout"), "InterestRatePayout"), "rateSpecification"), "FloatingRateSpecification")) or rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "payout"), "InterestRatePayout"), "rateSpecification"), "FixedRateSpecification"))), _then_fn2, _else_fn2)
    
    def _then_fn0():
        return SecurityFinanceCashSettlementAmount(rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "tradeState"), rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "date"), rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "quantity"), rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "payerReceiver"))
    
    def _else_fn0():
        return if_cond_fn(rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "payout"), "PerformancePayout")), _then_fn1, _else_fn1)
    
    payout = rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "payout")
    transfer =  if_cond_fn(rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "payout"), "AssetPayout")), _then_fn0, _else_fn0)
    transfer = _get_rosetta_object('Transfer', 'settlementDate', _get_rosetta_object('AdjustableOrAdjustedOrRelativeDate', 'adjustedDate', rosetta_resolve_attr(rosetta_resolve_attr(self, "instruction"), "date")))
    
    
    return transfer

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/event/common/functions/ResolveTransfer.java
package cdm.event.common.functions;

import cdm.base.math.Quantity;
import cdm.base.staticdata.party.PayerReceiver;
import cdm.event.common.CalculateTransferInstruction;
import cdm.event.common.Reset;
import cdm.event.common.TradeState;
import cdm.event.common.Transfer;
import cdm.event.common.Transfer.TransferBuilder;
import cdm.product.asset.FixedRateSpecification;
import cdm.product.asset.FloatingRateSpecification;
import cdm.product.asset.InterestRatePayout;
import cdm.product.asset.RateSpecification;
import cdm.product.template.AssetPayout;
import cdm.product.template.Payout;
import cdm.product.template.PerformancePayout;
import cdm.product.template.metafields.ReferenceWithMetaPayout;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.ModelObjectValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import com.rosetta.model.lib.records.Date;
import java.util.Optional;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(ResolveTransfer.ResolveTransferDefault.class)
public abstract class ResolveTransfer implements RosettaFunction {
	
	@Inject protected ModelObjectValidator objectValidator;
	
	// RosettaFunction dependencies
	//
	@Inject protected EquityCashSettlementAmount equityCashSettlementAmount;
	@Inject protected InterestCashSettlementAmount interestCashSettlementAmount;
	@Inject protected SecurityFinanceCashSettlementAmount securityFinanceCashSettlementAmount;

	/**
	* @param instruction 
	* @return transfer 
	*/
	public Transfer evaluate(CalculateTransferInstruction instruction) {
		Transfer.TransferBuilder transferBuilder = doEvaluate(instruction);
		
		final Transfer transfer;
		if (transferBuilder == null) {
			transfer = null;
		} else {
			transfer = transferBuilder.build();
			objectValidator.validate(Transfer.class, transfer);
		}
		
		return transfer;
	}

	protected abstract Transfer.TransferBuilder doEvaluate(CalculateTransferInstruction instruction);

	protected abstract MapperS<? extends ReferenceWithMetaPayout> payout(CalculateTransferInstruction instruction);

	public static class ResolveTransferDefault extends ResolveTransfer {
		@Override
		protected Transfer.TransferBuilder doEvaluate(CalculateTransferInstruction instruction) {
			Transfer.TransferBuilder transfer = Transfer.builder();
			return assignOutput(transfer, instruction);
		}
		
		protected Transfer.TransferBuilder assignOutput(Transfer.TransferBuilder transfer, CalculateTransferInstruction instruction) {
			if (exists(payout(instruction).<Payout>map("Type coercion", referenceWithMetaPayout0 -> referenceWithMetaPayout0 == null ? null : referenceWithMetaPayout0.getValue()).<AssetPayout>map("getAssetPayout", _payout -> _payout.getAssetPayout())).getOrDefault(false)) {
				transfer = toBuilder(securityFinanceCashSettlementAmount.evaluate(MapperS.of(instruction).<TradeState>map("getTradeState", calculateTransferInstruction -> calculateTransferInstruction.getTradeState()).get(), MapperS.of(instruction).<Date>map("getDate", calculateTransferInstruction -> calculateTransferInstruction.getDate()).get(), MapperS.of(instruction).<Quantity>map("getQuantity", calculateTransferInstruction -> calculateTransferInstruction.getQuantity()).get(), MapperS.of(instruction).<PayerReceiver>map("getPayerReceiver", calculateTransferInstruction -> calculateTransferInstruction.getPayerReceiver()).get()));
			} else if (exists(payout(instruction).<Payout>map("Type coercion", referenceWithMetaPayout1 -> referenceWithMetaPayout1 == null ? null : referenceWithMetaPayout1.getValue()).<PerformancePayout>map("getPerformancePayout", _payout -> _payout.getPerformancePayout())).getOrDefault(false)) {
				transfer = toBuilder(equityCashSettlementAmount.evaluate(MapperS.of(instruction).<TradeState>map("getTradeState", calculateTransferInstruction -> calculateTransferInstruction.getTradeState()).get(), MapperS.of(instruction).<Date>map("getDate", calculateTransferInstruction -> calculateTransferInstruction.getDate()).get()));
			} else if (exists(payout(instruction).<Payout>map("Type coercion", referenceWithMetaPayout2 -> referenceWithMetaPayout2 == null ? null : referenceWithMetaPayout2.getValue()).<InterestRatePayout>map("getInterestRatePayout", _payout -> _payout.getInterestRatePayout()).<RateSpecification>map("getRateSpecification", interestRatePayout -> interestRatePayout.getRateSpecification()).<FloatingRateSpecification>map("getFloatingRateSpecification", rateSpecification -> rateSpecification.getFloatingRateSpecification())).or(exists(payout(instruction).<Payout>map("Type coercion", referenceWithMetaPayout3 -> referenceWithMetaPayout3 == null ? null : referenceWithMetaPayout3.getValue()).<InterestRatePayout>map("getInterestRatePayout", _payout -> _payout.getInterestRatePayout()).<RateSpecification>map("getRateSpecification", interestRatePayout -> interestRatePayout.getRateSpecification()).<FixedRateSpecification>map("getFixedRateSpecification", rateSpecification -> rateSpecification.getFixedRateSpecification()))).getOrDefault(false)) {
				final ReferenceWithMetaPayout referenceWithMetaPayout4 = payout(instruction).get();
				transfer = toBuilder(interestCashSettlementAmount.evaluate(MapperS.of(instruction).<TradeState>map("getTradeState", calculateTransferInstruction -> calculateTransferInstruction.getTradeState()).get(), (referenceWithMetaPayout4 == null ? null : referenceWithMetaPayout4.getValue()), MapperS.of(instruction).<Reset>mapC("getResets", calculateTransferInstruction -> calculateTransferInstruction.getResets()).getMulti(), MapperS.of(instruction).<Date>map("getDate", calculateTransferInstruction -> calculateTransferInstruction.getDate()).get()));
			} else {
				transfer = null;
			}
			
			transfer
				.getOrCreateSettlementDate()
				.setAdjustedDateValue(MapperS.of(instruction).<Date>map("getDate", calculateTransferInstruction -> calculateTransferInstruction.getDate()).get());
			
			return Optional.ofNullable(transfer)
				.map(o -> o.prune())
				.orElse(null);
		}
		
		@Override
		protected MapperS<? extends ReferenceWithMetaPayout> payout(CalculateTransferInstruction instruction) {
			return MapperS.of(instruction).<ReferenceWithMetaPayout>map("getPayout", calculateTransferInstruction -> calculateTransferInstruction.getPayout());
		}
	}
}
